File: src-tauri/src/commands/gateway.rs
//! Usage: Gateway lifecycle / status / session / circuit commands.

use crate::app::gateway_service::{self, GatewayActiveSessionSummary};
use crate::app_state::{ensure_db_ready, DbInitState};
use crate::gateway_runtime_access::app_gateway_status;
use crate::{gateway, settings};

#[derive(Debug, Clone, serde::Deserialize, specta::Type)]
#[serde(rename_all = "camelCase")]
pub(crate) struct GatewayUpstreamProxyInput {
    proxy_url: String,
    proxy_username: Option<String>,
    proxy_password: Option<String>,
}

fn resolve_gateway_upstream_proxy_context(
    app: &tauri::AppHandle,
    input: &GatewayUpstreamProxyInput,
) -> Result<
    (
        Option<String>,
        crate::gateway::http_client::GatewaySelfCheckContext,
    ),
    String,
> {
    let cfg = settings::read(app).map_err(|err| err.to_string())?;
    let proxy_url = gateway::http_client::build_effective_proxy_url(
        Some(input.proxy_url.as_str()),
        input.proxy_username.as_deref(),
        input.proxy_password.as_deref(),
    )?;
    let context = gateway::http_client::self_check_context_from_settings(&cfg)
        .map_err(|err| err.to_string())?;
    Ok((proxy_url, context))
}

#[tauri::command]
#[specta::specta]
pub(crate) fn gateway_status(app: tauri::AppHandle) -> gateway::GatewayStatus {
    app_gateway_status(&app)
}

#[tauri::command]
#[specta::specta]
pub(crate) async fn gateway_check_port_available(
    app: tauri::AppHandle,
    port: u16,
) -> Result<bool, String> {
    gateway_service::check_port_available(app, port)
        .await
        .map_err(Into::into)
}

#[tauri::command]
#[specta::specta]
pub(crate) async fn gateway_sessions_list(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    limit: Option<u32>,
) -> Result<Vec<GatewayActiveSessionSummary>, String> {
    let db = ensure_db_ready(app.clone(), db_state.inner()).await?;
    gateway_service::list_active_sessions(app, db, limit)
        .await
        .map_err(Into::into)
}

#[tauri::command]
#[specta::specta]
pub(crate) async fn gateway_circuit_status(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    cli_key: String,
) -> Result<Vec<gateway::GatewayProviderCircuitStatus>, String> {
    let db = ensure_db_ready(app.clone(), db_state.inner()).await?;
    gateway_service::circuit_status(app, db, cli_key)
        .await
        .map_err(Into::into)
}

#[tauri::command]
#[specta::specta]
pub(crate) async fn gateway_circuit_reset_provider(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    provider_id: i64,
) -> Result<bool, String> {
    let db = ensure_db_ready(app.clone(), db_state.inner()).await?;
    gateway_service::circuit_reset_provider(app, db, provider_id)
        .await
        .map_err(Into::into)
}

#[tauri::command]
#[specta::specta]
pub(crate) async fn gateway_circuit_reset_cli(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    cli_key: String,
) -> Result<usize, String> {
    let db = ensure_db_ready(app.clone(), db_state.inner()).await?;
    gateway_service::circuit_reset_cli(app, db, cli_key)
        .await
        .map_err(Into::into)
}

#[tauri::command]
#[specta::specta]
pub(crate) async fn gateway_start(
    app: tauri::AppHandle,
    db_state: tauri::State<'_, DbInitState>,
    preferred_port: Option<u16>,
) -> Result<gateway::GatewayStatus, String> {
    let db = ensure_db_ready(app.clone(), db_state.inner()).await?;
    gateway_service::start_and_sync(app, db, preferred_port)
        .await
        .map_err(Into::into)
}

#[tauri::command]
#[specta::specta]
pub(crate) async fn gateway_stop(app: tauri::AppHandle) -> Result<gateway::GatewayStatus, String> {
    gateway_service::stop_and_restore(app)
        .await
        .map_err(Into::into)
}

#[tauri::command]
#[specta::specta]
pub(crate) fn gateway_upstream_proxy_validate(
    app: tauri::AppHandle,
    input: GatewayUpstreamProxyInput,
) -> Result<(), String> {
    let (proxy_url, context) = resolve_gateway_upstream_proxy_context(&app, &input)?;
    gateway::http_client::validate_proxy_with_context(proxy_url.as_deref(), &context)
}

#[tauri::command]
#[specta::specta]
pub(crate) async fn gateway_upstream_proxy_test(
    app: tauri::AppHandle,
    input: GatewayUpstreamProxyInput,
) -> Result<(), String> {
    let (proxy_url, context) = resolve_gateway_upstream_proxy_context(&app, &input)?;
    gateway::http_client::test_proxy_with_context(proxy_url.as_deref(), &context).await
}

#[tauri::command]
#[specta::specta]
pub(crate) async fn gateway_upstream_proxy_detect_ip(
    app: tauri::AppHandle,
    input: GatewayUpstreamProxyInput,
) -> Result<String, String> {
    let (proxy_url, context) = resolve_gateway_upstream_proxy_context(&app, &input)?;
    gateway::http_client::detect_proxy_exit_ip_with_context(proxy_url.as_deref(), &context).await
}

File: src/services/gateway/gateway.ts
import {
  commands,
  type GatewayActiveSessionSummary,
  type GatewayUpstreamProxyInput,
  type GatewayProviderCircuitStatus,
  type GatewayStatus,
} from "../../generated/bindings";
import { invokeGeneratedIpc, type GeneratedCommandResult } from "../generatedIpc";

export type { GatewayProviderCircuitStatus, GatewayStatus };
export type GatewayActiveSession = GatewayActiveSessionSummary;

export async function gatewayStatus() {
  return invokeGeneratedIpc<GatewayStatus>({
    title: "获取网关状态失败",
    cmd: "gateway_status",
    invoke: () => commands.gatewayStatus(),
  });
}

export async function gatewayStart(preferredPort?: number) {
  return invokeGeneratedIpc<GatewayStatus>({
    title: "启动网关失败",
    cmd: "gateway_start",
    args: { preferredPort: preferredPort ?? null },
    invoke: () =>
      commands.gatewayStart(preferredPort ?? null) as Promise<GeneratedCommandResult<GatewayStatus>>,
  });
}

export async function gatewayStop() {
  return invokeGeneratedIpc<GatewayStatus>({
    title: "停止网关失败",
    cmd: "gateway_stop",
    invoke: () => commands.gatewayStop() as Promise<GeneratedCommandResult<GatewayStatus>>,
  });
}

export async function gatewayCheckPortAvailable(port: number) {
  return invokeGeneratedIpc<boolean>({
    title: "检查端口可用性失败",
    cmd: "gateway_check_port_available",
    args: { port },
    invoke: () =>
      commands.gatewayCheckPortAvailable(port) as Promise<GeneratedCommandResult<boolean>>,
  });
}

export async function gatewaySessionsList(limit?: number) {
  return invokeGeneratedIpc<GatewayActiveSession[]>({
    title: "获取会话列表失败",
    cmd: "gateway_sessions_list",
    args: { limit: limit ?? null },
    invoke: () =>
      commands.gatewaySessionsList(limit ?? null) as Promise<
        GeneratedCommandResult<GatewayActiveSession[]>
      >,
  });
}

export async function gatewayCircuitStatus(cliKey: string) {
  return invokeGeneratedIpc<GatewayProviderCircuitStatus[]>({
    title: "获取熔断器状态失败",
    cmd: "gateway_circuit_status",
    args: { cliKey },
    invoke: () =>
      commands.gatewayCircuitStatus(cliKey) as Promise<
        GeneratedCommandResult<GatewayProviderCircuitStatus[]>
      >,
  });
}

export async function gatewayCircuitResetProvider(providerId: number) {
  return invokeGeneratedIpc<boolean>({
    title: "重置 Provider 熔断器失败",
    cmd: "gateway_circuit_reset_provider",
    args: { providerId },
    invoke: () =>
      commands.gatewayCircuitResetProvider(providerId) as Promise<GeneratedCommandResult<boolean>>,
  });
}

export async function gatewayCircuitResetCli(cliKey: string) {
  return invokeGeneratedIpc<number>({
    title: "重置 CLI 熔断器失败",
    cmd: "gateway_circuit_reset_cli",
    args: { cliKey },
    invoke: () =>
      commands.gatewayCircuitResetCli(cliKey) as Promise<GeneratedCommandResult<number>>,
  });
}

type GatewayUpstreamProxyAuthInput = {
  proxyUrl: string;
  proxyUsername?: string;
  proxyPassword?: string;
};

function buildGatewayUpstreamProxyInput({
  proxyUrl,
  proxyUsername,
  proxyPassword,
}: GatewayUpstreamProxyAuthInput): GatewayUpstreamProxyInput {
  return {
    proxyUrl,
    proxyUsername: proxyUsername ?? null,
    proxyPassword: proxyPassword ?? null,
  };
}

export async function gatewayUpstreamProxyValidate({
  proxyUrl,
  proxyUsername,
  proxyPassword,
}: GatewayUpstreamProxyAuthInput) {
  const input = buildGatewayUpstreamProxyInput({ proxyUrl, proxyUsername, proxyPassword });
  return invokeGeneratedIpc<null, null>({
    title: "代理配置验证失败",
    cmd: "gateway_upstream_proxy_validate",
    args: { input },
    invoke: () =>
      commands.gatewayUpstreamProxyValidate(input) as Promise<GeneratedCommandResult<null>>,
    nullResultBehavior: "return_fallback",
    fallback: null,
  });
}

export async function gatewayUpstreamProxyTest({
  proxyUrl,
  proxyUsername,
  proxyPassword,
}: GatewayUpstreamProxyAuthInput) {
  const input = buildGatewayUpstreamProxyInput({ proxyUrl, proxyUsername, proxyPassword });
  return invokeGeneratedIpc<null, null>({
    title: "代理连接测试失败",
    cmd: "gateway_upstream_proxy_test",
    args: { input },
    invoke: () =>
      commands.gatewayUpstreamProxyTest(input) as Promise<GeneratedCommandResult<null>>,
    nullResultBehavior: "return_fallback",
    fallback: null,
  });
}

export async function gatewayUpstreamProxyDetectIp({
  proxyUrl,
  proxyUsername,
  proxyPassword,
}: GatewayUpstreamProxyAuthInput) {
  const input = buildGatewayUpstreamProxyInput({ proxyUrl, proxyUsername, proxyPassword });
  return invokeGeneratedIpc<string>({
    title: "代理出口 IP 检测失败",
    cmd: "gateway_upstream_proxy_detect_ip",
    args: { input },
    invoke: () =>
      commands.gatewayUpstreamProxyDetectIp(input) as Promise<GeneratedCommandResult<string>>,
  });
}
